openpype/pipeline/delivery.py

```diff
@@ -162,8 +164,41 @@ def deliver_single_file(
     if not os.path.exists(delivery_folder):
         os.makedirs(delivery_folder)
 
-    log.debug("Copying single: {} -> {}".format(src_path, delivery_path))
-    _copy_file(src_path, delivery_path)
+    if has_renumbered_frame:
+        src_dir = os.path.dirname(src_path)
+        src_filepaths = os.listdir(src_dir)
+        src_collections, remainders = clique.assemble(src_filepaths)
+
+        src_collection = src_collections[0]
+        src_indexes = list(src_collection.indexes)
+
+        destination_indexes = [
+            renumber_frame + idx
+            for idx in range(len(src_indexes))
+        ]
+
+        dst_filepaths = []
+        for index in destination_indexes:
```

Comment:
The idea here is to only modify the frame on the filename, right?

So if we have `beep_boop.1001.jpeg` we want to get `beep_boop.0001.jpeg` , correct?

IF the assumption above is correct, we could then do:
```py
collections, remainder = clique.assemble(["beep_boop.1001.jpg", "beep_boop.1002.jpg"])
source_collection = collections[0]
# <Collection "beep_boop.%d.jpg [1001-1002]">
destination_indexes = [
    renumber_frame + idx
    for idx in range(len(col.indexes))
]

destination_collection = clique.Collection(
    source_collection.head,
    source_collection.tail,
    source_collection.padding, # Unsure about this one,
    indexes=destination_indexes
)
# <Collection "beep_boop.%d.jpg [0-1]">
```

Reply:
I am actually talking about other things. It will probably resolve if we are using representation data if a single frame has been published(if i dont get it wrong, it will publish in `{filename}.{frame}.{ext}` and the file transmission from integrate files will remove the frame if it is a single frame). It will probably break the delivery action if we still using `os.listdir(src_dir)` with single frame when the renumber option enabled.

Reply:
Regarding @Minkiu 's example I don't think that's the perfect future-compatible fix either.  think at some point we'll need to support publishing sequences with holes in it - instead of force closing the gaps. However I do really like his code style, it's much more readable like that. 

I think this could be a quite a decent readable solution:

```python
def shift_collection(collection, start_offset, relative=True):
    """Shift frames of a clique.Collection.
    
    When relative is True `start_offset` will be an offset from
    the current start frame of the collection shifting it by `start_offset`.
    When relative is False `start_offset` will be the new start
    frame of the sequence - shifting the rest of the frames along.
    
    Arguments:
        collection (clique.Collection): Input collection.
        start_offset (int): Offset to apply (or start frame to set if
            relative is False)
        relative (bool): Whether the start offset is relative or the
            the absolute new start frame.
    
    Returns:
        clique.Collection: Shifted collection
    
    """
    first_frame = next(iter(col.indexes))
    shift = first_frame - start_offset if not relative else start_offset
    return clique.Collection(
        head=source_collection.head, 
        tail=source_collection.tail, 
        padding=source_collection.padding, 
        indexes={idx + shift for idx in collection.indexes} 
    )
```

Which means that if you have any renumbering you can just do:
```python
if has_shift:
    dst_collection = shift_collection(dst_collection, start_offset=start_offset, relative=relative)
```

Right?

Reply:
I assume this function is prepare for the future use as there is `relative=True` as argument?

Reply:
I thought this delivery action had a way to set the start frame explicitly OR shift the start frame. 

```
So input sequence is e.g. frame 1001-1100. 
Then renumber could either shift it by 100 (relative): 1101-1200
Or renumber it to a new start frame (absolute), e.g. frame 1: 0001-0100.
```

_Note:_ It's good to be a aware that if the input sequence is 1001-1100 that `clique.Collection` itself doesn't recognize a padding of the source sequence, so it'll have padding `0` and the output does becomes e.g. `1-100` not `0001-0100`.

If you want to have at least the padding width of the start frame of the source collection you'd still need to do:
```python
first_frame = next(iter(src_collection.indexes))
dst_collection.padding = len(str(first_frame))
```